File: data/ai_model/src/features.rs
/// WAF feature extraction — exact port of ml_waf/features.py.
/// 29 dense float32 features per HTTP request string.
///
/// Input format (multi-line):
///   Line 0:  "METHOD /url [body]"
///   Lines 1+: "Header: value"   (User-Agent, Cookie, Referer, etc.)
///
/// Old single-line format "METHOD /url body" is still accepted.
use once_cell::sync::Lazy;
use regex::Regex;

pub const NUM_FEATURES: usize = 29;

/// Mirror build_clean_dataset.py truncation limits so live inference matches training.
const MAX_URL_BYTES: usize = 4_096;
const MAX_BODY_BYTES: usize = 8_192;

/// Truncate `s` to at most `max_bytes`, respecting UTF-8 char boundaries.
fn truncate(s: &str, max_bytes: usize) -> &str {
    if s.len() <= max_bytes {
        return s;
    }
    let mut idx = max_bytes;
    while !s.is_char_boundary(idx) {
        idx -= 1;
    }
    &s[..idx]
}

pub const FEATURE_NAMES: [&str; NUM_FEATURES] = [
    "request_len",
    "method_id",
    "path_len",
    "query_len",
    "body_len",
    "num_params",
    "entropy",
    "digit_ratio",
    "upper_ratio",
    "special_char_count",
    "single_quote_count",
    "double_quote_count",
    "angle_bracket_count",
    "semicolon_count",
    "pct_encoded_count",
    "sql_keyword_count",
    "xss_pattern_count",
    "path_traversal_count",
    "cmd_injection_count",
    "scanner_count",
    "ssrf_count",
    "php_pattern_count",
    "null_byte_count",
    "hex_encode_count",
    "crlf_inject_count",
    "double_encode_count",
    "ssti_count",
    "header_count",
    "header_entropy",
];

static SQL_RE: Lazy<Regex> = Lazy::new(|| {
    Regex::new(r"(?i)\b(select|union|insert|update|delete|drop|create|alter|exec|execute|where|from|having|order|group|join|table|database|schema|char|nchar|varchar|cast|convert|declare|waitfor|xp_|sp_|0x)\b").unwrap()
});

// Context gate for sql_keyword_count (parity with ml_waf/features.py `_SQL_CTX`).
static SQL_CTX_RE: Lazy<Regex> = Lazy::new(|| {
    Regex::new(r#"(?is)union\s+(?:all\s+)?select|\bselect\b[^a-zA-Z]*?(?:\*|@@|count\s*\(|distinct\b|top\s+\d|[\w`\[\]]+\s*,)|\bselect\s+(?:version|substring|substr|concat|char|count|current_user|current_database|current_setting|session_user|user|database|group_concat|load_file)\b|@@\w+|\bfrom\s+(?:information_schema|mysql\.|pg_|sys\.|sysobjects)|\binsert\s+into\b|\bdelete\s+from\b|\bupdate\b[^;]{0,80}?\bset\b|\bdrop\s+(?:table|database|schema)\b|\balter\s+table\b|\bcreate\s+(?:table|database)\b|\b(?:order|group)\s+by\s+\d|\bhaving\s+\d|\binto\s+(?:outfile|dumpfile)\b|\bwaitfor\s+delay\b|\b(?:information_schema|load_file|extractvalue|updatexml|benchmark|sleep|pg_sleep)\s*\(|\b(?:xp_|sp_)\w+|(?:--|\#)\s*$|/\*.*?\*/|;\s*(?:select|insert|update|delete|drop|union|create|alter)\b|['"]\s*(?:or|and)\s|\b(?:or|and)\b\s*['"\d][^=<>]{0,12}?[=<>]\s*['"\d\w(]"#).unwrap()
});

static XSS_RE: Lazy<Regex> = Lazy::new(|| {
    Regex::new(r"(?i)(<script|javascript:|vbscript:|onload=|onerror=|onclick=|onfocus=|alert\(|confirm\(|prompt\(|document\.cookie|document\.write|eval\(|<iframe|<img\s|<svg|srcdoc=)").unwrap()
});

static SCANNER_RE: Lazy<Regex> = Lazy::new(|| {
    Regex::new(r"(?i)(nikto|sqlmap|nmap|masscan|acunetix|nessus|openvas|dirbuster|gobuster|wfuzz|w3af|commix)").unwrap()
});

static PCT_RE: Lazy<Regex> = Lazy::new(|| Regex::new(r"%[0-9a-fA-F]{2}").unwrap());

static CMD_RE: Lazy<Regex> =
    Lazy::new(|| Regex::new(r"\||&&|\$\(|`[^`]*`").unwrap());

static SSRF_RE: Lazy<Regex> = Lazy::new(|| {
    Regex::new(r"(?i)(?:127\.0\.0\.1|localhost|169\.254\.|0\.0\.0\.0|::1|file://|dict://|gopher://|ftp://)").unwrap()
});

static PHP_RE: Lazy<Regex> = Lazy::new(|| {
    Regex::new(r"(?i)(?:\.php|eval\(|base64_decode\(|system\(|passthru\(|shell_exec\(|phpinfo\(|\$_(?:GET|POST|REQUEST|FILES)\[)").unwrap()
});

static NULL_BYTE_RE: Lazy<Regex> =
    Lazy::new(|| Regex::new(r"%00|\\x00|\\u0000").unwrap());

static HEX_ENCODE_RE: Lazy<Regex> =
    Lazy::new(|| Regex::new(r"0x[0-9a-fA-F]{4,}").unwrap());

static CRLF_RE: Lazy<Regex> =
    Lazy::new(|| Regex::new(r"%0[aAdD]|\\r\\n|\r\n").unwrap());

static DBL_ENC_RE: Lazy<Regex> =
    Lazy::new(|| Regex::new(r"%25[0-9a-fA-F]{2}").unwrap());

static SSTI_RE: Lazy<Regex> = Lazy::new(|| {
    Regex::new(
        r"(?is)(\{\{.*?\}\})|(\$\{[^}]{1,200}\})|(#\{[^}]{1,200}\})|(<%=.*?%>)|(\?\s*new\s*\()|(__(class|mro|subclasses|globals|builtins|import)__)|(freemarker\.template|velocity\.tools)|(\{\s*\d+\s*\*\s*\d+\s*\})"
    ).unwrap()
});

fn method_id(method: &str) -> f32 {
    match method.to_ascii_uppercase().as_str() {
        "GET"     => 0.0,
        "POST"    => 1.0,
        "PUT"     => 2.0,
        "DELETE"  => 3.0,
        "PATCH"   => 4.0,
        "HEAD"    => 5.0,
        "OPTIONS" => 6.0,
        _         => 7.0,
    }
}

/// Decode %XX percent-encoding — mirrors Python's `urllib.parse.unquote`.
///
/// Collects decoded bytes first, then interprets them as UTF-8 (with lossy
/// fallback for invalid sequences).  The old byte-by-byte char-push approach
/// produced Latin-1 characters for multi-byte UTF-8 sequences:
///   %E4%B8%AD → 'ä' + '¸' + soft-hyphen  (wrong)
/// vs Python:
///   %E4%B8%AD → '中'                       (correct)
///
/// For pure ASCII inputs (the common WAF case) both approaches are identical.
fn url_decode(s: &str) -> String {
    let bytes = s.as_bytes();
    let mut raw: Vec<u8> = Vec::with_capacity(s.len());
    let mut i = 0;
    while i < bytes.len() {
        if bytes[i] == b'%' && i + 2 < bytes.len() {
            let hi = (bytes[i + 1] as char).to_digit(16);
            let lo = (bytes[i + 2] as char).to_digit(16);
            if let (Some(h), Some(l)) = (hi, lo) {
                raw.push((h * 16 + l) as u8);
                i += 3;
                continue;
            }
        }
        raw.push(bytes[i]);
        i += 1;
    }
    // Interpret as UTF-8; replace invalid sequences (same as Python's errors='replace').
    String::from_utf8_lossy(&raw).into_owned()
}

/// Shannon entropy — mirrors Python's `_entropy()` exactly.
///
/// Uses a HashMap keyed by char so every Unicode code point gets its own
/// bucket.  The old fixed-array approach (256 slots) silently merged all
/// chars with code point > 255 into a single bucket, underestimating
/// entropy for non-ASCII payloads.
fn shannon_entropy(s: &str) -> f32 {
    if s.is_empty() {
        return 0.0;
    }
    let mut counts: std::collections::HashMap<char, u32> = std::collections::HashMap::new();
    let mut n = 0u32;
    for c in s.chars() {
        *counts.entry(c).or_insert(0) += 1;
        n += 1;
    }
    let n = n as f32;
    counts
        .values()
        .map(|&c| {
            let p = c as f32 / n;
            -p * p.log2()
        })
        .sum()
}

/// Extract 29 WAF features from a raw HTTP request string.
///
/// Accepts both formats:
///   Single-line : "METHOD /url body"
///   Multi-line  : "METHOD /url body\nUser-Agent: ...\nCookie: ..."
pub fn extract_features(request: &str) -> [f32; NUM_FEATURES] {
    // Split into the request line and optional header lines.
    let (first_line, headers_text): (&str, String) = match request.find('\n') {
        Some(nl) => (&request[..nl], request[nl + 1..].replace('\n', " ")),
        None     => (request, String::new()),
    };

    let mut parts = first_line.splitn(3, ' ');
    let method = parts.next().unwrap_or("GET");
    let url    = truncate(parts.next().unwrap_or("/"), MAX_URL_BYTES);
    let body   = truncate(parts.next().unwrap_or(""), MAX_BODY_BYTES);

    let (path, query) = match url.find('?') {
        Some(pos) => (&url[..pos], &url[pos + 1..]),
        None      => (url, ""),
    };

    // `full` includes url + body + header values so every regex pattern
    // also scans User-Agent, Cookie, Referer, etc. — mirrors Python logic.
    // Separate full strings:
    // - full_for_chars: URL + body only (for character counts, no header noise)
    // - full_for_patterns: URL + body + headers (for regex-based attack detection in all fields)
    let full_for_chars = {
        let mut s = url.to_string();
        if !body.is_empty() { s.push(' '); s.push_str(body); }
        s
    };

    let full_for_patterns = {
        let mut s = url.to_string();
        if !body.is_empty()         { s.push(' '); s.push_str(body); }
        if !headers_text.is_empty() { s.push(' '); s.push_str(&headers_text); }
        s
    };

    let full_dec_for_patterns = url_decode(&full_for_patterns);
    let n = full_for_chars.chars().count().max(1) as f32;

    // num_params counts query-string and body params, not header values.
    let num_params =
        (if query.is_empty() { 0 } else { query.matches('&').count() + 1 })
        + (if body.is_empty()  { 0 } else { body.matches('&').count() + 1 });

    let digit_count  = full_for_chars.chars().filter(|c| c.is_ascii_digit()).count() as f32;
    let upper_count  = full_for_chars.chars().filter(|c| c.is_ascii_uppercase()).count() as f32;
    let special_count = full_for_chars
        .chars()
        .filter(|c| matches!(c, '\'' | '"' | '<' | '>' | ';' | '=' | '%' | '&' | '+'))
        .count() as f32;

    // "../" is already lowercase — no need to lowercase full_dec before matching.
    let path_traversal = full_dec_for_patterns.matches("../").count() as f32;

    [
        // Python len() counts Unicode *characters*; Rust str::len() counts UTF-8
        // *bytes*.  Use chars().count() so the values match for non-ASCII input.
        request.chars().count() as f32,                                    // 0  total length in chars
        method_id(method),                                                 // 1
        path.chars().count() as f32,                                       // 2
        query.chars().count() as f32,                                      // 3
        body.chars().count() as f32,                                       // 4  body length in chars
        num_params as f32,                                                 // 5  query + body params
        shannon_entropy(&full_for_chars),                                  // 6
        digit_count / n,                                                   // 7
        upper_count / n,                                                   // 8
        special_count,                                                     // 9
        full_for_chars.matches('\'').count() as f32,                       // 10
        full_for_chars.matches('"').count() as f32,                        // 11
        (full_for_chars.matches('<').count() + full_for_chars.matches('>').count()) as f32,    // 12
        full_for_chars.matches(';').count() as f32,                        // 13
        PCT_RE.find_iter(&full_for_chars).count() as f32,                  // 14 raw
        if SQL_CTX_RE.is_match(&full_dec_for_patterns) {                   // 15 decoded, context-gated
            SQL_RE.find_iter(&full_dec_for_patterns).count() as f32
        } else {
            0.0
        },
        XSS_RE.find_iter(&full_dec_for_patterns).count() as f32,           // 16 decoded
        path_traversal,                                                    // 17 decoded
        CMD_RE.find_iter(&full_dec_for_patterns).count() as f32,           // 18 decoded
        SCANNER_RE.find_iter(&full_dec_for_patterns).count() as f32,       // 19 decoded
        SSRF_RE.find_iter(&full_dec_for_patterns).count() as f32,          // 20 decoded
        PHP_RE.find_iter(&full_dec_for_patterns).count() as f32,           // 21 decoded
        NULL_BYTE_RE.find_iter(&full_for_patterns).count() as f32,          // 22 raw — %00 vanishes after url_decode, must scan raw string
        HEX_ENCODE_RE.find_iter(&full_for_chars).count() as f32,           // 23 raw
        CRLF_RE.find_iter(&full_for_chars).count() as f32,                 // 24 raw
        DBL_ENC_RE.find_iter(&full_for_chars).count() as f32,              // 25 raw
        SSTI_RE.find_iter(&full_dec_for_patterns).count() as f32,          // 26 decoded
        request.matches('\n').count() as f32,                              // 27 header_count
        shannon_entropy(&headers_text),                                    // 28 header_entropy
    ]
}

File: data/ml_waf/features.py
"""
WAF feature extraction — identical logic ported to Rust (waf_infer/src/features.rs).
29 dense engineered features per HTTP request.
"""
import math
import re
import urllib.parse
from collections import Counter

FEATURE_NAMES = [
    "request_len",       # 0  total byte length
    "method_id",         # 1  GET=0 POST=1 PUT=2 DELETE=3 PATCH=4 HEAD=5 OPTIONS=6 other=7
    "path_len",          # 2  length of URL path
    "query_len",         # 3  length of query string
    "body_len",          # 4  length of request body
    "num_params",        # 5  number of key=value params (query + body)
    "entropy",           # 6  Shannon entropy of url+body
    "digit_ratio",       # 7  fraction of digit chars
    "upper_ratio",       # 8  fraction of uppercase chars
    "special_char_count",  # 9  count of  ' " < > ; = % & +
    "single_quote_count",  # 10
    "double_quote_count",  # 11
    "angle_bracket_count", # 12 '<' + '>'
    "semicolon_count",     # 13
    "pct_encoded_count",   # 14 %XX sequences
    "sql_keyword_count",   # 15
    "xss_pattern_count",   # 16
    "path_traversal_count",# 17 occurrences of '../'
    "cmd_injection_count", # 18
    "scanner_count",       # 19
    "ssrf_count",          # 20 SSRF indicators
    "php_pattern_count",   # 21 PHP-specific patterns
    "null_byte_count",     # 22 null byte injection
    "hex_encode_count",    # 23 hex-encoded payloads
    "crlf_inject_count",   # 24 CRLF / header injection
    "double_encode_count", # 25 double URL encoding (%25xx)
    "ssti_count",          # 26 Server-Side Template Injection patterns
    "header_count",        # 27 number of HTTP headers present
    "header_entropy",      # 28 Shannon entropy of header text (high for real browsers, 0 for bare probes)
]

NUM_FEATURES = len(FEATURE_NAMES)

_SQL = re.compile(
    r"\b(select|union|insert|update|delete|drop|create|alter|exec|execute|"
    r"where|from|having|order|group|join|table|database|schema|"
    r"char|nchar|varchar|cast|convert|declare|waitfor|xp_|sp_|0x)\b",
    re.IGNORECASE,
)
_XSS = re.compile(
    r"(<script|javascript:|vbscript:|onload=|onerror=|onclick=|onfocus=|"
    r"alert\(|confirm\(|prompt\(|document\.cookie|document\.write|eval\(|"
    r"<iframe|<img\s|<svg|srcdoc=)",
    re.IGNORECASE,
)
_SCANNER = re.compile(
    r"(nikto|sqlmap|nmap|masscan|acunetix|nessus|openvas|"
    r"dirbuster|gobuster|wfuzz|w3af|commix)",
    re.IGNORECASE,
)
_PCT = re.compile(r"%[0-9a-fA-F]{2}")
# `\(\s*\)\s*\{` = shellshock function-definition signature "() {"
# (CVE-2014-6271), delivered via headers like Accept-Language / User-Agent.
_CMD = re.compile(r"\||&&|\$\(|`[^`]*`|\(\s*\)\s*\{")
_SSRF = re.compile(
    r"(?:127\.0\.0\.1|localhost|169\.254\.|0\.0\.0\.0|::1|"
    r"file://|dict://|gopher://|ftp://)",
    re.IGNORECASE,
)
_PHP = re.compile(
    r"(?:\.php|eval\(|base64_decode\(|system\(|passthru\(|"
    r"shell_exec\(|phpinfo\(|\$_(?:GET|POST|REQUEST|FILES)\[)",
    re.IGNORECASE,
)
_NULL_BYTE = re.compile(r"%00|\\x00|\\u0000")
_HEX_ENCODE = re.compile(r"0x[0-9a-fA-F]{4,}")
_CRLF = re.compile(r"%0[aAdD]|\\r\\n|\r\n")
_DBL_ENC = re.compile(r"%25[0-9a-fA-F]{2}")
_SSTI = re.compile(
    r"(\{\{.*?\}\})"                                      # Jinja2/Twig: {{expr}}
    r"|(\$\{[^}]{1,200}\})"                               # FreeMarker/Spring EL: ${expr}
    r"|(#\{[^}]{1,200}\})"                                # Spring EL: #{expr}
    r"|(<%=.*?%>)"                                        # ERB/JSP: <%= expr %>
    r"|(\?\s*new\s*\()"                                   # FreeMarker: ?new()
    r"|(__(class|mro|subclasses|globals|builtins|import)__)"  # Python dunders
    r"|(freemarker\.template|velocity\.tools)"             # template engine namespaces
    r"|(\{\s*\d+\s*\*\s*\d+\s*\})",                      # {N*N} arithmetic probe
    re.IGNORECASE | re.DOTALL,
)

_METHOD_MAP = {
    "GET": 0, "POST": 1, "PUT": 2, "DELETE": 3,
    "PATCH": 4, "HEAD": 5, "OPTIONS": 6,
}


def _entropy(s: str) -> float:
    if not s:
        return 0.0
    counts = Counter(s)
    n = len(s)
    return -sum((v / n) * math.log2(v / n) for v in counts.values())


def extract_features(request: str) -> list[float]:
    # Format: "METHOD /url [body]\nHeader: val\n..."
    # Old single-line format is still handled (no \n present).
    nl = request.find("\n")
    if nl == -1:
        first_line, headers_text = request, ""
    else:
        first_line = request[:nl]
        headers_text = request[nl + 1:].replace("\n", " ")

    parts = first_line.split(" ", 2)
    method = parts[0] if parts else "GET"
    url    = parts[1] if len(parts) > 1 else "/"
    body   = parts[2] if len(parts) > 2 else ""

    path, _, query = url.partition("?")

    num_params = (query.count("&") + 1 if query else 0) + (
        body.count("&") + 1 if body else 0
    )

    full_for_chars = url + (" " + body if body else "")
    full_for_patterns = url + (" " + body if body else "") + (" " + headers_text if headers_text else "")
    n = max(len(full_for_chars), 1)

    try:
        full_dec_for_patterns = urllib.parse.unquote(full_for_patterns)
    except Exception:
        full_dec_for_patterns = full_for_patterns

    return [
        # ── length / structural (raw) ────────────────────────────────────────
        float(len(request)),                                    # 0
        float(_METHOD_MAP.get(method.upper(), 7)),              # 1
        float(len(path)),                                       # 2
        float(len(query)),                                      # 3
        float(len(body)),                                       # 4
        float(num_params),                                      # 5
        _entropy(full_for_chars),                               # 6
        sum(1 for c in full_for_chars if c.isdigit()) / n,      # 7
        sum(1 for c in full_for_chars if c.isupper()) / n,      # 8
        float(sum(1 for c in full_for_chars if c in "'\"><;=%&+")),  # 9
        float(full_for_chars.count("'")),                       # 10
        float(full_for_chars.count('"')),                       # 11
        float(full_for_chars.count("<") + full_for_chars.count(">")),  # 12
        float(full_for_chars.count(";")),                       # 13
        # ── encoding indicators (raw — the encoding itself is the signal) ───
        float(len(_PCT.findall(full_for_chars))),               # 14
        # ── attack patterns (decoded — catch encoded payloads in URL+body+headers) ───
        float(len(_SQL.findall(full_dec_for_patterns))),        # 15
        float(len(_XSS.findall(full_dec_for_patterns))),        # 16
        float(full_dec_for_patterns.lower().count("../")),      # 17
        float(len(_CMD.findall(full_dec_for_patterns))),        # 18
        float(len(_SCANNER.findall(full_dec_for_patterns))),    # 19
        float(len(_SSRF.findall(full_dec_for_patterns))),       # 20
        float(len(_PHP.findall(full_dec_for_patterns))),        # 21
        float(len(_NULL_BYTE.findall(full_for_patterns))),       # 22 raw — %00 vanishes after unquote; scan raw to catch it in Cookie/headers
        float(len(_HEX_ENCODE.findall(full_for_chars))),        # 23
        float(len(_CRLF.findall(full_for_chars))),              # 24
        float(len(_DBL_ENC.findall(full_for_chars))),           # 25
        float(len(_SSTI.findall(full_dec_for_patterns))),       # 26
        float(request.count("\n")),                              # 27 header_count
        _entropy(headers_text),                                  # 28 header_entropy
    ]
